### ISSUE TITLE
Works with React Native?

### ISSUE THREAD

**Issue by ffxsam** [bitcoinjs/bitcoinjs-lib#582]
Does anyone know if this library will work with React Native? Theoretically any JS could should, but I wanted to know if it would operate securely, or if there were any caveats.

**Comment by dcousens** [comment]
That is a good question.   I don't have the answer for you,  maybe @jprichardson does.
However,  the steps to testing would be:
* Attempt to run the test suite in React Native
* Investigate React Native's secure random number generation source

If the results of those two steps are satisfactory,  then I think you'd be perfectly OK to use this library :+1:

**Comment by jprichardson** [comment]
> I don't have the answer for you, maybe @jprichardson does.

Unfortunately I don't either, I haven't been able to use React Native yet 😞  However @mvayngrib may know.

**Comment by mvayngrib** [comment]
bitcoinjs-lib works fine in react native. Currently you'll need [rn-nodeify](https://github.com/mvayngrib/rn-nodeify), at least until [this issue](https://productpains.com/post/react-native/packager-support-resolvealias-ala-webpack/) is resolved

**Comment by dcousens** [comment]
@mvayngrib that is interesting that they are different.

**Comment by mvayngrib** [comment]
@dcousens yep, javascript only has async access to native functions

**Comment by dcousens** [comment]
@ffxsam save yourself reviewing code,  pre-fetch the RNG data of SecRandomCopyBytes and pass that in instead.

pseudo-code:
``` javascript
native.SecRandomCopyBytes(32, function (randomData) {
  // WARNING: could throw if the 32 bytes aren't adequate,  so you may have to repeat this process if it throws
  var keyPair = ECPair.makeRandom({
    rng: function () {
      return randomData
    }
  })

  // ...
})
```

**Comment by mvayngrib** [comment]
you'll need a piece of native code to make that work, but react-native-randombytes basically does those five lines for you:
js: https://github.com/mvayngrib/react-native-randombytes/blob/master/index.js#L39
objc: https://github.com/mvayngrib/react-native-randombytes/blob/master/RNRandomBytes.m

**Comment by ffxsam** [comment]
This thread is a gold mine. Thanks!

**Comment by dcousens** [comment]
ping @39otrebla,  were you having issues with this?

**Comment by 39otrebla** [comment]
@dcousens yeah had some issues, randomBytes does not work on ReactNative out of the box, as well as node-dependent packages.. I did need some workarounds, but now it works perfectly

**Comment by dcousens** [comment]
@39otrebla can you think of a way we could incorporate support into this library without you needing to maintain a separate fork?

**Comment by dcousens** [comment]
@39otrebla any reason why you didn't just pass in your own `rng` parameter?
Or did `randombytes` as a package fail to import entirely?

**Comment by dcousens** [comment]
@fanatid if we used `require('buffer')` around the place,  that and people using the `rng:` parameter would mean we could claim this library is `react-native` compatible.

Thoughts?
It'd probably require a few changes upstream to all our dependencies,  and even `crypto-browserify`... what other downsides would there be?

**Comment by dcousens** [comment]
@jprichardson @mvayngrib @calvinmetcalf @dominictarr @feross @fanatid

tl;dr: If we all used `require('buffer')` around the place we could claim this library,  and many others,  as `react-native` compatible.

Thoughts?

It would require quite a few changes upstream to various dependencies.

I know this problem shouldn't be our responsibility,  but at the same time it is just a nasty side effect of nodes implicit "non-standard" JS globals.

We took the approach in https://github.com/crypto-browserify to use `require('create-hash')` instead of `require('crypto')` as a way to reduce `require` bloat.

Maybe we can take the same initiative here (except in allowing for just a plug-and-play approach for new tools such as `react-native`).

**Comment by mvayngrib** [comment]
@dcousens i think the initiative mostly belongs on the react-native packager end, to enable shimming / resolve-aliasing a la browserify and webpack. I currently use bitcoinjs-lib and other libs with [rn-nodeify](https://github.com/mvayngrib/rn-nodeify)'s hackery, and a bunch of native shims i co-developed with other react native fans. @philikon has developed [https://github.com/philikon/ReactNativify](ReactNativify), which is a healthier approach, but one I haven't tested yet. I suggested to him recently to try it out on bitcoinjs-lib and he said it worked: https://github.com/mvayngrib/rn-nodeify/issues/25#issuecomment-277853745

also see: https://productpains.com/post/react-native/packager-support-resolvealias-ala-webpack

**Comment by dominictarr** [comment]
I'd be happy to merge a pr adding explicit `require('buffer')` to any of my modules, though I think there may be a lot of them, and a better way would be if

**Comment by dcousens** [comment]
@dominictarr 
> and a better way would be if

if? :smiley:

**Comment by dominictarr** [comment]
sorry ... if a pull request was made to react-native to support `Buffer`

**Comment by dcousens** [comment]
@dominictarr I'm thinking of re-visiting this one by and just adding `require('buffer')`...

**Comment by dominictarr** [comment]
@dcousens lets do it

**Comment by dcousens** [comment]
Closing in favour of https://github.com/bitcoinjs/bitcoinjs-lib/issues/797
